cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java

```diff
@@ -259,8 +346,17 @@ public OperationSnapshot createCluster(String name, String imageVersion, Map<Str
       } else {
         clusterConfig.setNetworkUri(network);
       }
-      for (String targetTag : getFirewallTargetTags()) {
-        clusterConfig.addTags(targetTag);
+
+      // if internal ip is being used firewall rules might not exists and is not needed
+      if (!useInternalIP) {
+        for (String targetTag : getFirewallTargetTags()) {
+          clusterConfig.addTags(targetTag);
+        }
+      }
+
+      // if internal ip is prefered then create dataproc cluster without external ip for better security
+      if (useInternalIP) {
```

Comment:
Combine the two `if` to a `if-else`.

Reply:
Removed the other if as the firewall checks needs to be done for private instances too